File: 0506. Relative Ranks/0506.py
# time complexity: O(nlogn)
# space complexity: O(n)
class Solution(object):
    def findRelativeRanks(self, score):
        """
        :type score: List[int]
        :rtype: List[str]
        """
        # create an empty array of length n
        n = len(score)
        result = [""] * n

        # put each score in a list as (index, score)
        indexed_scores = []
        for i in range(n):
            indexed_scores.append((i, score[i]))

        # sort scores by the second element (x[1]) of each element in descending order
        # x = (0, 10) → x[1] = 10
        # x = (1, 3) → x[1] = 3
        indexed_scores.sort(key=lambda x: x[1], reverse=True)

        # medal names for top 3
        medals = ["Gold Medal", "Silver Medal", "Bronze Medal"]

        # assign ranks
        for rank in range(n):
            index = indexed_scores[rank][0]  # original index
            if rank < 3:
                result[index] = medals[rank]
            else:
                result[index] = str(rank + 1)
            
        return result

File: 0506. Relative Ranks/0506.java
// time complexity: O(n^2)
// space complexity: O(n)
class Solution {
    public String[] findRelativeRanks(int[] score) {
        // all the elements of score are unique
        // label the elements as their ranks.
        int n = score.length;

        // an array to store ranks
        int[] arr = new int[n];

        // calculate the ranks
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (score[i] > score[j])
                    arr[j]++;
            }
        }

        // result array for rank strings
        String[] results = new String[n];

        // assign medals or rank numbers based on position
        for (int i = 0; i < n; i++) {
            if (arr[i] == 0)
                results[i] = "Gold Medal";
            else if (arr[i] == 1)
                results[i] = "Silver Medal";
            else if (arr[i] == 2)
                results[i] = "Bronze Medal";
            else
                results[i] = String.valueOf(arr[i] + 1); // add 1 because ranks start from 1
        }

        return results;
    }
}